core/imds/src/main/java/software/amazon/awssdk/imds/internal/DefaultEc2Metadata.java

```diff
@@ -86,191 +88,110 @@ public Ec2Metadata.Builder toBuilder() {
-    }
-
-    @Override
-    public String toString() {
-        return "DefaultEc2Metadata{" +
-               "retries=" + retryPolicy +
-               ", endpoint='" + endpoint + '\'' +
-               ", tokenTtl=" + tokenTtl +
-               ", endpointMode='" + endpointMode + '\'' +
-               ", httpDebugOutput='" + httpDebugOutput + '\'' +
-               ", httpClient= " + httpClient.clientName() +
-               '}';
-    }
-
     /**
      * Gets the specified instance metadata value by the given path.
      * @param path  Input path
+     * @throws UncheckedIOException If an IOException occurs during transfer.
+     * @throws SdkClientException if the request for a token or the request for the Metadata does not have a 2XX SUCCESS response,
+     *                            if the maximum number of retries is reached,
+     *                            or if another unknown problem occurred.
      * @return Instance metadata value as part of MetadataResponse Object
      */
     @Override
     public MetadataResponse get(String path) {
-
-        MetadataResponse metadataResponse = null;
-        AbortableInputStream abortableInputStream = null;
-
-        for (int tries = 1 ; tries <= retryPolicy.numRetries() + 1; tries ++) {
-
+        Throwable lastCause = null;
+        for (int tries = 0; tries < retryPolicy.numRetries(); tries++) {
+            RetryPolicyContext retryPolicyContext = RetryPolicyContext.builder().retriesAttempted(tries).build();
             try {
-                Optional<String> token = getToken();
-                if (token.isPresent()) {
-                    HttpExecuteResponse response = getDataHttpResponse(path, token.get());
-
-                    int statusCode = response.httpResponse().statusCode();
-                    Optional<AbortableInputStream> responseBody = response.responseBody();
-
-                    if (statusCode == 200) {
-                        if (!responseBody.isPresent()) {
-                            throw SdkClientException.builder()
-                                                     .message("Response body empty with Status Code "  + statusCode).build();
-                        }
-                        abortableInputStream = responseBody.get();
-                        String data = IoUtils.toUtf8String(abortableInputStream);
-                        metadataResponse = new MetadataResponse(data);
-                        return metadataResponse;
-                    }
-                    handleException(statusCode, path);
-                }
-                //TODO Create IMDS Custom Exception
-            } catch (IOException io) {
-                log.warn(() -> "Received an IOException ", io);
-            } finally {
-                IoUtils.closeQuietly(abortableInputStream, log.logger());
+                String token = getToken();
+                return doRequest(path, token);
+            } catch (SdkClientException sdkClientException) {
```

Comment:
I think in previous code, only IOException is retryable, or at least it's what I think it should be. It seems we are retrying both exceptions now.

Reply:
So no retry attempts should be done on `SdkClientException`, gotcha. It make sense since the request has almost no chances of success on the next try.

Reply:
Yeah, I'm surprised that tests are still passing with this change. I think we should add some tests for this. It's probably out of scope of this PR, could you create a backlog item for it?